examples/Results/main.roc

```diff
@@ -0,0 +1,106 @@
+app [main] {
+    pf: platform "https://github.com/roc-lang/basic-cli/releases/download/0.12.0/Lb8EgiejTUzbggO2HVVuPJFkwvvsfW6LojkLR20kTVE.tar.br",
+}
+
+import pf.Stdout
+import pf.Task exposing [Task]
+
+# This function parses a string formatted like "{FirstName} {LastName} was born in {Year}"
+# This is the most verbose version. It's a bit too nested, we will do better below.
+parsePersonVerbose = \line ->
+    when line |> Str.split " was born in " is
+        [fullName, birthYearStr] ->
+            when fullName |> Str.split " " is
+                [firstName, lastName] ->
+                    when Str.toU16 birthYearStr is
+                        Ok birthYear ->
+                            Ok { firstName, lastName, birthYear }
+
+                        Err _ -> Err (ParsingErr InvalidBirthYearFormat)
+
+                _ -> Err (ParsingErr InvalidNameFormat)
+
+        _ -> Err (ParsingErr InvalidRecordFormat)
+
+## This is a utility function, used below to convert [a, b] to Result (a, b)
+toPair = \listWithTwoItems ->
+    when listWithTwoItems is
+        [a, b] -> Ok (a, b)
+        _ -> Err NotAPair
+
+## This is a much nicer parser, using the backpassing syntax.
+## See https://github.com/roc-lang/roc/blob/main/roc-for-elm-programmers.md#backpassing
+## This version returns "raw" errors (i.e., `Err NotAPair` or `Err InvalidNumStr`).
+parsePersonBackpassing = \line ->
+    (fullName, birthYearStr) <- line |> Str.split " was born in " |> toPair |> Result.try
+    (firstName, lastName) <- fullName |> Str.split " " |> toPair |> Result.try
+    birthYear <- Str.toU16 birthYearStr |> Result.try
+    Ok { firstName, lastName, birthYear }
+
+## This function is the same as the previous one, except it uses `Result.mapErr` to
+## return more informative errors, just like `parsePersonVerbose`.
+parsePersonBackpassingV2 = \line ->
+    (fullName, birthYearStr) <- line
+        |> Str.split " was born in "
+        |> toPair
+        |> Result.mapErr (\_ -> Err (ParsingErr InvalidRecordFormat))
```

Comment:
I would remove `ParsingErr` from all of these errors, I don't think it is idiomatic.

If a user is directly handling the error, it is just noise.
If a user is not directly handling they error, they can easily decide to wrap it:
`parePersonBackpassingV2 line |> Result.mapErr ParsingErr`

Oh, also, you are adding a redundant `Err`. I think this would result in a `Err (Err (ParsingErr ...))` on failure.

Reply:
Good points, thanks! Fixed.